File: android/app/src/main/kotlin/dev/agentdeck/terrarium/TerrariumConfig.kt
package dev.agentdeck.terrarium

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

/** Color palette and timing constants for the terrarium scene. */
object TerrariumColors {
    // Background layers
    val DeepSea = Color(0xFF0A1628)
    val MidWater = Color(0xFF0F2744)
    val ShallowWater = Color(0xFF163B5C)

    // Claude Code mascot (pixel art — matching official terracotta)
    val ClaudeBody = Color(0xFFC07058)       // muted terracotta/copper
    val ClaudeBodyLight = Color(0xFFD08870)  // THINKING pulse bright
    val ClaudeBodyDark = Color(0xFFA05840)   // shadow
    val ClaudeEye = Color(0xFF2D1F16)        // dark brown

    // Crayfish (OpenClaw brand: #FF4D4D→#991B1B gradient, #00E5CC teal eyes)
    val CrayfishShell = Color(0xFFFF4D4D)
    val CrayfishDark = Color(0xFF991B1B)
    val CrayfishClaw = Color(0xFFFF4D4D)
    val CrayfishEye = Color(0xFF00E5CC)
    val CrayfishBodyLight = Color(0xFFFF6B6B)  // ROUTING pulse bright

    // Neon Tetra
    val TetraNeon = Color(0xFF00E5FF)
    val TetraBody = Color(0xFF1E40AF)
    val TetraFin = Color(0xFFFF6B6B)
    val TetraStripe = Color(0xFF00E5FF)

    // Environment
    val BubbleWhite = Color(0x40FFFFFF)
    val BubbleHighlight = Color(0x80FFFFFF)
    val CausticsLight = Color(0x14FFFFFF)
    val SandBase = Color(0xFF2A1F14)
    val SandLight = Color(0xFF3D2E1F)
    val RockDark = Color(0xFF1A1A2E)
    val RockMid = Color(0xFF2D2D44)
    val RockLight = Color(0xFF3A3A55)
    val KelpGreen = Color(0xFF22C55E)
    val KelpDark = Color(0xFF166534)

    // LED cables
    val LEDGreen = Color(0xFF22C55E)
    val LEDAmber = Color(0xFFFBBF24)
    val LEDRed = Color(0xFFEF4444)

    // Holographic UI
    val HoloBlue = Color(0x6000E5FF)
    val HoloText = Color(0xB000E5FF)

    // Error state
    val ErrorTint = Color(0x40EF4444)

    // HUD overlay
    val HUDBg = Color(0x80000000)
    val HUDText = Color(0xFFE2E8F0)
    val HUDSubtext = Color(0xFF94A3B8)
}

/** Layout and sizing constants. */
object TerrariumLayout {
    // Scene proportions (fraction of canvas)
    const val SAND_HEIGHT_FRACTION = 0.18f
    const val ROCK_HEIGHT_FRACTION = 0.25f
    const val WATER_SURFACE_Y_FRACTION = 0.05f

    // Octopus sizing (fraction of canvas width)
    const val OCTOPUS_BODY_RADIUS_FRACTION = 0.07f
    const val OCTOPUS_CENTER_X_FRACTION = 0.4f
    const val OCTOPUS_CENTER_Y_FRACTION = 0.45f
    // (TENTACLE_LENGTH_FRACTION removed — pixel mascot has no tentacles)

    // Crayfish sizing
    const val CRAYFISH_WIDTH_FRACTION = 0.14f
    const val CRAYFISH_CENTER_X_FRACTION = 0.78f
    const val CRAYFISH_CENTER_Y_FRACTION = 0.75f

    // Tetra sizing
    const val TETRA_SIZE_FRACTION = 0.015f
    const val TETRA_COUNT = 7
}

/** Animation timing constants. */
object TerrariumTiming {
    // Octopus
    const val FLOAT_PERIOD_MS = 4000f
    const val FLOAT_AMPLITUDE_FRACTION = 0.015f
    const val TENTACLE_WAVE_SPEED = 2.5f
    const val THINKING_PULSE_SPEED = 3.0f
    const val TYPING_SPEED = 8.0f

    // Crayfish
    const val CLAW_CLAP_PERIOD_MS = 1200f
    const val EYE_FLASH_PERIOD_MS = 800f

    // Tetra
    const val BOID_SPEED = 0.3f
    const val STREAM_SPEED = 1.5f
    const val SEPARATION_RADIUS = 0.04f
    const val ALIGNMENT_RADIUS = 0.08f
    const val COHESION_RADIUS = 0.12f

    // Bubbles
    const val BUBBLE_RISE_SPEED = 0.08f
    const val BUBBLE_WOBBLE_SPEED = 3.0f
    const val CALM_SPAWN_INTERVAL_MS = 2000f
    const val ACTIVE_SPAWN_INTERVAL_MS = 300f
    const val ERROR_SPAWN_INTERVAL_MS = 100f

    // Environment
    const val CAUSTICS_SPEED = 1.5f
    const val KELP_SWAY_SPEED = 1.0f
    const val LED_PULSE_SPEED = 2.0f

    // Transitions
    const val STATE_TRANSITION_MS = 500f
    const val EINK_WIPE_MS = 300f
    const val EINK_DEBOUNCE_MS = 500L
}

File: apple/AgentDeck/Terrarium/TerrariumConfig.swift
// TerrariumConfig.swift — Colors, layout constants, timing
// Ported from android TerrariumConfig.kt

import SwiftUI

// MARK: - Environment Visual State

enum EnvironmentVisualState {
    case dark    // Disconnected — dim/off
    case calm    // Idle — gentle caustics, slow bubbles
    case active  // Processing — bright caustics, more bubbles
    case alert   // Awaiting input — pulsing highlights
}

// MARK: - Colors

enum TerrariumColors {
    // Background layers
    static let deepSea = Color(red: 0.039, green: 0.086, blue: 0.157)         // #0A1628
    static let midWater = Color(red: 0.059, green: 0.153, blue: 0.267)        // #0F2744
    static let shallowWater = Color(red: 0.086, green: 0.231, blue: 0.361)    // #163B5C

    // Claude Code (Octopus)
    static let claudeBody = Color(red: 0.753, green: 0.439, blue: 0.345)      // #C07058
    static let claudeBodyLight = Color(red: 0.816, green: 0.533, blue: 0.439) // #D08870
    static let claudeBodyDark = Color(red: 0.627, green: 0.345, blue: 0.251)  // #A05840
    static let claudeEye = Color(red: 0.176, green: 0.122, blue: 0.086)       // #2D1F16

    // Crayfish
    static let crayfishShell = Color(red: 1.0, green: 0.302, blue: 0.302)     // #FF4D4D
    static let crayfishDark = Color(red: 0.6, green: 0.106, blue: 0.106)      // #991B1B
    static let crayfishClaw = Color(red: 1.0, green: 0.302, blue: 0.302)      // #FF4D4D
    static let crayfishEye = Color(red: 0.0, green: 0.898, blue: 0.8)         // #00E5CC
    static let crayfishBodyLight = Color(red: 1.0, green: 0.42, blue: 0.42)   // #FF6B6B

    // Neon Tetra
    static let tetraNeon = Color(red: 0.0, green: 0.898, blue: 1.0)           // #00E5FF
    static let tetraBody = Color(red: 0.118, green: 0.251, blue: 0.686)       // #1E40AF
    static let tetraFin = Color(red: 1.0, green: 0.42, blue: 0.42)            // #FF6B6B
    static let tetraStripe = Color(red: 0.0, green: 0.898, blue: 1.0)         // #00E5FF

    // Jellyfish (Codex CLI) — matches icon gradient: lavender/pink top → vivid blue bottom
    static let jellyfishBell = Color(red: 0.380, green: 0.400, blue: 0.880)      // #6166E0 mid blue-indigo
    static let jellyfishDeep = Color(red: 0.200, green: 0.260, blue: 0.780)      // #3342C7 vivid deep blue
    static let jellyfishHighlight = Color(red: 0.700, green: 0.580, blue: 0.900) // #B394E5 lavender-pink (top glow)
    static let jellyfishGlow = Color(red: 0.560, green: 0.580, blue: 0.950)      // #8F94F2 periwinkle
    static let jellyfishTentacle = Color(red: 0.450, green: 0.490, blue: 0.920)  // #737DEA mid
    static let jellyfishNameBg = Color(red: 0.333, green: 0.380, blue: 0.878).opacity(0.6)

    // Environment
    static let bubbleWhite = Color.white.opacity(0.25)      // 0x40FFFFFF
    static let bubbleHighlight = Color.white.opacity(0.5)    // 0x80FFFFFF
    static let causticsLight = Color.white.opacity(0.08)     // 0x14FFFFFF
    static let sandBase = Color(red: 0.165, green: 0.122, blue: 0.078)        // #2A1F14
    static let sandLight = Color(red: 0.239, green: 0.180, blue: 0.122)       // #3D2E1F
    static let rockDark = Color(red: 0.102, green: 0.102, blue: 0.180)        // #1A1A2E
    static let rockMid = Color(red: 0.176, green: 0.176, blue: 0.267)         // #2D2D44
    static let rockLight = Color(red: 0.227, green: 0.227, blue: 0.333)       // #3A3A55
    static let kelpGreen = Color(red: 0.133, green: 0.773, blue: 0.369)       // #22C55E
    static let kelpDark = Color(red: 0.086, green: 0.396, blue: 0.204)        // #166534

    // LED cables
    static let ledGreen = Color(red: 0.133, green: 0.773, blue: 0.369)        // #22C55E
    static let ledAmber = Color(red: 0.984, green: 0.749, blue: 0.141)        // #FBBF24
    static let ledRed = Color(red: 0.937, green: 0.267, blue: 0.267)          // #EF4444

    // Holographic UI
    static let holoBlue = Color(red: 0.0, green: 0.898, blue: 1.0).opacity(0.375)
    static let holoText = Color(red: 0.0, green: 0.898, blue: 1.0).opacity(0.69)

    // Error state
    static let errorTint = Color(red: 0.937, green: 0.267, blue: 0.267).opacity(0.25)

    // HUD overlay
    static let hudBg = Color.black.opacity(0.5)
    static let hudText = Color(red: 0.886, green: 0.910, blue: 0.878)         // #E2E8F0
    static let hudSubtext = Color(red: 0.580, green: 0.639, blue: 0.722)      // #94A3B8

    // Sick/desaturated crayfish
    static let crayfishSick = Color(red: 0.545, green: 0.482, blue: 0.482) // #8B7B7B

    // Food crumbs
    static let foodCyan = Color(red: 0.0, green: 0.898, blue: 1.0)            // tools
    static let foodAmber = Color(red: 0.984, green: 0.749, blue: 0.141)       // messages
    static let foodGreen = Color(red: 0.133, green: 0.773, blue: 0.369)       // code

    // Name tag
    static let claudeNameBg = Color(red: 0.753, green: 0.439, blue: 0.345).opacity(0.6)

    /// Linear interpolation between two SwiftUI Colors using resolved RGBA components.
    static func lerpColor(_ a: Color, _ b: Color, _ t: Float) -> Color {
        let f = min(1, max(0, t))
        if f < 0.01 { return a }
        if f > 0.99 { return b }
        let env = EnvironmentValues()
        let ra = a.resolve(in: env)
        let rb = b.resolve(in: env)
        return Color(
            red: Double(ra.red * (1 - f) + rb.red * f),
            green: Double(ra.green * (1 - f) + rb.green * f),
            blue: Double(ra.blue * (1 - f) + rb.blue * f),
            opacity: Double(ra.opacity * (1 - f) + rb.opacity * f)
        )
    }
}

// MARK: - Layout

enum TerrariumLayout {
    // Scene proportions (fraction of canvas)
    static let sandHeightFraction: Float = 0.35
    static let rockHeightFraction: Float = 0.25

    // Octopus
    static let octopusBodyRadius: Float = 0.050
    static let pixelAspect: Float = 2.0
    static let pixelGap: Float = 0.5

    // Standing positions
    static let standingY: Float = 0.62
    static let standingYDeep: Float = 0.75
    static let jitterRange: Float = 0.03

    // Swim bounds (octopus)
    static let swimMinX: Float = 0.18
    static let swimMaxX: Float = 0.55
    static let swimMinY: Float = 0.15
    static let swimMaxY: Float = 0.55

    // Swim bounds (tetra)
    static let tetraMinX: Float = 0.03
    static let tetraMaxX: Float = 0.92
    static let tetraMinY: Float = 0.08
    static let tetraMaxY: Float = 0.61

    // Crayfish
    static let crayfishWidthFraction: Float = 0.11
    static let crayfishDefaultX: Float = 0.78
    static let crayfishSittingY: Float = 0.64
    static let crayfishRoutingY: Float = 0.42

    // Jellyfish
    static let jellyfishBodyRadius: Float = 0.050
    static let jellyfishDefaultX: Float = 0.50
    static let jellyfishIdleY: Float = 0.45
    static let jellyfishProcessingY: Float = 0.20

    // Surface
    static let surfaceY: Float = 0.04
}

// MARK: - Timing

enum TerrariumTiming {
    // Octopus
    static let bobPeriod: Float = 4.0
    static let bobAmplitude: Float = 0.015
    static let swayAmplitude: Float = 0.008
    static let thinkingPulseSpeed: Float = 3.0

    // Tentacle
    static let tentacleSpeedWorking: Float = 1.5
    static let tentacleSpeedFloating: Float = 0.8
    static let tentacleSpeedAsking: Float = 1.0
    static let tentacleAmpWorking: Float = 0.08
    static let tentacleAmpFloating: Float = 0.04
    static let tentacleAmpAsking: Float = 0.05

    // Arm bob
    static let armSpeedWorking: Float = 1.0
    static let armSpeedFloating: Float = 0.5
    static let armAmpWorking: Float = 0.06
    static let armAmpFloating: Float = 0.02

    // Starburst
    static let starburstArmCount = 10
    static let starburstRotSpeed: Float = 0.5
    static let starburstPulseSpeed: Float = 3.0

    // Swimming waypoints
    static let swimLerpRate: Float = 3.0
    static let waypointMinInterval: Float = 1.5
    static let waypointMaxInterval: Float = 3.0

    // Crayfish
    static let heartbeatPeriod: Float = 4.0
    static let clawClapPeriod: Float = 1.2
    static let eyeFlashPeriod: Float = 0.8

    // Bubble
    static let bubbleRiseSpeed: Float = 0.08
    static let bubbleWobbleSpeed: Float = 3.0
    static let bubbleMaxCount = 70
    static let bubbleSpawnCalm: Float = 2.0      // seconds
    static let bubbleSpawnActive: Float = 0.3
    static let bubbleSpawnAlert: Float = 0.45

    // Environment
    static let causticsSpeed: Float = 1.5
    static let kelpSwaySpeed: Float = 1.0
    static let ledPulseSpeed: Float = 2.0

    // Creature bubble exhales
    static let octoBubbleInterval: Float = 2.5
    static let crayfishBubbleInterval: Float = 1.5

    // Tetra
    static let tetraSize: Float = 0.015
    static let tetraTailSpeed: Float = 8.0
    static let schoolAttractorWeight: Float = 0.4
    static let separationRadius: Float = 0.04
    static let alignmentRadius: Float = 0.08
    static let cohesionRadius: Float = 0.12
    static let maxFood = 30
    static let foodLifetime: Float = 5.0
    static let foodEatRadius: Float = 0.03
}
